[UC01#3] Designed the importation logic for JSON files and import them

Matte0Sacc0 opened the pull request:
This code implements the task #3 and allows the JSON files containing the Cards information to be read and imported into the application for each Card Game present.
It's the base feature to allow the task #4

Matte0Sacc0 committed:
Implemented all the tests in ImportJSONTest for the method importJSON in the class ListenerImpl
And also implemented the parser to convert strings in JsonObjects

alessioarcara reviewed (changes requested):
Hi matteo,

There are few things missing here:
- [UC01#3] in pull request title
- Link to task #3 in development
- Resolve conflicts in s01_UC01#3_Import-Cards-From-JSON before approving pull request

Regarding the code:

```java
public String[]parseJSONString(String jsonString) {
     jsonString = jsonString.substring(1, jsonString.length() - 1);
     if ( jsonString.charAt(jsonString.length() - 1) == ',' ) {
         jsonString = jsonString.substring(0, jsonString.length() - 1);
     }
    
     String[] parsedJSON = jsonString.split("\\},\\{");
     for (int i = 0; i < parsedJSON.length; i++) {
         if (parsedJSON[i].charAt(0) != '{') {
             parsedJSON[i] = "{" + parsedJSON[I];
         }
         if (parsedJSON[i].charAt(parsedJSON[i].length() - 1) != '}') {
             parsedJSON[i] += "}";
         }
     }
     return parsedJSON;
}

/*
* this method serves the scope of importing information about the cards of the three games from JSON files
* the inputs taken are :
*  - jsonString : the string that represents the array of json objects to be imported
*  - type       : the type of deck we want to import
*                 - y : yugi-oh
*                 - m : magic
*                 - p : pokemon
*/
public CardDecorator[] importJSON(String jsonString, char type) {
     Gson gson = new Gson();
    
     String[] inputJSON = parseJSONString(jsonString);
    
     CardDecorator[] cards = new CardDecorator[inputJSON.length];
    
      // common fields between the 3 types of deck
     switch (type) {
         case 'y' :
         {
             cards = new YuGiOhCardDecorator[inputJSON.length];
             for (int i = 0; i < inputJSON.length; i++) {
                 JsonObject json = gson.fromJson(inputJSON[i], JsonObject.class);
    
                 // fields
                 String name = json.has("name") ? json.get("name").getAsString() : "";
                 String description = json.has("desc") ? json.get("desc").getAsString() : "";
                 String types = json.has("type") ? json.get("type").getAsString() : "";
                 String race = json.has("race") ? json.get("race").getAsString() : "";
                 String imageUrl = json.has("image_url") ? json.get("image_url").getAsString() : "";
                 String smallImageUrl = json.has("small_image_url") ? json.get("small_image_url").getAsString() : "";
    
                 cards[i] = new YuGiOhCardDecorator(new CardImpl(name, description, types), race, imageUrl, smallImageUrl);
             }
             break;
         }
         case 'm' :
         {
             cards = new MagicCardDecorator[inputJSON.length];
             for (int i = 0; i < inputJSON.length; i++) {
                 JsonObject json = gson.fromJson(inputJSON[i], JsonObject.class);
    
                 // fields
                 String name = json.has("name") ? json.get("name").getAsString() : "";
                 String description = json.has("text") ? json.get("text").getAsString() : "";
                 String types = json.has("types") ? json.get("types").getAsString() : "";
                 String artist = json.has("artist") ? json.get("artist").getAsString() : "";
                 String rarity = json.has("rarity") ? json.get("rarity").getAsString() : "";
                 boolean hasFoil = (json.has("hasFoil")) && (json.get("hasFoil").getAsInt() != 0);
                 boolean isAlternative = (json.has("isAlternative")) && (json.get("isAlternative").getAsInt() != 0);
                 boolean isFullArt = (json.has("isFullArt")) && (json.get("isFullArt").getAsInt() != 0);
                 boolean isPromo = (json.has("isPromo")) && (json.get("isPromo").getAsInt() != 0);
                 boolean isReprint = (json.has("isReprint")) && (json.get("isReprint").getAsInt() != 0);
    
                 cards[i] = new MagicCardDecorator(new CardImpl(name, description, types),
                         artist, rarity,
                         hasFoil, isAlternative, isFullArt, isPromo, isReprint);
             }
             break;
         }
         case 'p' :
         {
             cards = new PokemonCardDecorator[inputJSON.length];
             for (int i = 0; i < inputJSON.length; i++) {
                 JsonObject json = gson.fromJson(inputJSON[i], JsonObject.class);
    
                 // fields
                 String name = json.has("name") ? json.get("name").getAsString() : "";
                 String description = json.has("description") ? json.get("description").getAsString() : "";
                 String types = json.has("types") ? json.get("types").getAsString() : "";
                 String illustrator = json.has("illustrator") ? json.get("illustrator").getAsString() : "";
                 String image = json.has("image") ? json.get("image").getAsString() : "";
                 String rarity = json.has("rarity") ? json.get("rarity").getAsString() : "";
    
                 JsonObject variants = json.has("variants") ? json.getAsJsonObject("variants") : null;
                 boolean isFirstEdition = false;
                 boolean isHolo = false;
                 boolean isNormal = false;
                 boolean isReverse = false;
                 boolean isPromo = false;
    
                 if (json.has("variants")){
                     isFirstEdition = variants.has("firstEdition") && variants.get("firstEdition").getAsBoolean();
                     isHolo = variants.has("holo") && variants.get("holo").getAsBoolean();
                     isNormal = variants.has("normal") && variants.get("normal").getAsBoolean();
                     isReverse = variants.has("reverse") && variants.get("reverse").getAsBoolean();
                     isPromo = variants.has("wPromo") && variants.get("wPromo").getAsBoolean();
                 }
    
    
                 cards[i] = new PokemonCardDecorator(new CardImpl(name, description, types),
                         illustrator, image, rarity,
                         isFirstEdition, isHolo, isNormal, isReverse, isPromo);
             }
             break;
         }
         default : return cards;
     }
    
      return cards;
}
```

- The code is tightly coupled with a specific JSON format;
- It has hard-coded values, if these values will change in future, this code need to be updated;
- Also it is not very readable;
- There is no error handling for cases where the JSON is malformed.

Even though it's not great, it can work for the JSON we need to load.

Would be a good move to:
1) Add error handling for cases where the JSON is malformed
2) Refactor code responsible to parse and import JSON to its own class

may be a good candidate as a non-functional enhancement:
- Refactor with `strategy pattern`

Matte0Sacc0 committed:
Created JsonReader file and moved in all the methods from ListenerImpl.java

Handled all the MalformedJsonException manually in the Parser created before and caught all their throws

Modeled Strategy Pattern inside JsonParser class. The different strategies haven't been split in classes because of their static

Matte0Sacc0 committed:
Refactoring codice
Implementazione pattern strategy e lettura json attraverso fileReader

Cons: non è più possibile fare parse delle stringe formattate come json (ma non è necessario in quesot tipo di applicazione)

Validazione con test

Matte0Sacc0 committed:
Merge branch 'development' into s01_UC01#3_Import-Cards-From-JSON

# Conflicts:
#	src/main/java/com/aadm/cardexchange/server/ListenerImpl.java
#	src/main/java/com/aadm/cardexchange/shared/models/MagicCardDecorator.java
#	src/test/java/com/aadm/cardexchange/CardExchangeSuite.java

Matte0Sacc0 committed:
Refactoring codice
Implementazione pattern strategy e lettura json attraverso fileReader

Cons: non è più possibile fare parse delle stringe formattate come json (ma non è necessario in quesot tipo di applicazione)

Validazione con test

Merge conflitti con avanzamento branch developement

alessioarcara reviewed (approved):
I really appreciated the refactors you made to eliminate duplicate code and the JSONParser is well implemented and adheres to the open-closed principle using the strategy pattern. Good work!
